onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -430,14 +464,17 @@ Status CUDAExecutionProvider::OnRunEnd(bool sync_stream, const onnxruntime::RunO
 }
 
 bool CUDAExecutionProvider::IsGraphCaptureEnabled() const {
-  return info_.enable_cuda_graph;
+  return info_.enable_cuda_graph && GetPerThreadContext().IsGraphCaptureAllowedOnRun();
 }
 
-bool CUDAExecutionProvider::IsGraphCaptured() const {
+bool CUDAExecutionProvider::IsGraphCaptured(int graph_annotation_id) const {
+  ORT_ENFORCE(GetPerThreadContext().SetCudaGraphAnnotationId(graph_annotation_id) == Status::OK());
   return GetPerThreadContext().IsGraphCaptured();
 }
 
-Status CUDAExecutionProvider::ReplayGraph() {
+Status CUDAExecutionProvider::ReplayGraph(int graph_annotation_id) {
+  // SetAnnotationId is not needed here
+  ORT_RETURN_IF_ERROR(GetPerThreadContext().SetCudaGraphAnnotationId(graph_annotation_id));
   return GetPerThreadContext().ReplayGraph();
```

Comment:
Shall we change interface like GetPerThreadContext().ReplayGraph(graph_annotation_id)?

Reply:
I think we can keep the GetPerThreadContext().ReplayGraph() and GetPerThreadContext().IsGraphCaptured() interfaces unchanged by using the SetCudaGraphAnnotationId() method